File: p090m/subsets_with_dup.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function (nums) {
  const ans = [];
  nums.sort();

  const backtrack = (wip, start) => {
    ans.push(Array.from(wip));
    for (let i = start; i < nums.length; ++i) {
      if (i > start && nums[i] === nums[i - 1]) continue;
      wip.push(nums[i]);
      backtrack(wip, i + 1);
      wip.pop();
    }
  };

  backtrack([], 0);
  return ans;
};
// TEST
[
  [
    [1, 2, 2],
    [[], [1], [1, 2], [1, 2, 2], [2], [2, 2]],
  ],
  [[0], [[], [0]]],
].forEach(([nums, expected]) => {
  const actual = subsetsWithDup(nums);
  console.log('Subsets with dups of', nums, '->', actual);
  console.assert(actual.toString() === expected.toString());
});

File: p090m/subsets_with_dup.py
from typing import List


class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums_sorted = sorted(nums)

        def backtrack(wip: List[int], start: int) -> None:
            ans.append(list(wip))
            for i in range(start, len(nums)):
                if i > start and nums_sorted[i] == nums_sorted[i - 1]:
                    continue
                wip.append(nums_sorted[i])
                backtrack(wip, i + 1)
                wip.pop()

        backtrack([], 0)
        return ans


# TESTS
for nums, expected in [
    ([1, 2, 2], [[], [1], [1, 2], [1, 2, 2], [2], [2, 2]]),
    ([0], [[], [0]]),
]:
    sol = Solution()
    actual = sol.subsetsWithDup(nums)
    print("Subsets with dups of", nums, "->", actual)
    assert actual == expected
